### PR TITLE
fix(facets): Fix facets service using only the first filter entity to…

### PR DESCRIPTION
**Pull request by javieri-empathy**
Fixes an issue where only the first filter was used to create an entity when updating the facets. If the first filter had a modifier, or was from a "less common" type, all the filters from different facets were updated using this behavior. 

For example, in added regression test, we are only setting the `gender` facet to single select, but because it is the first facet in the list of facets that we pass to the `updateFacets` method, the `size` facet was using also this behavior wrongly, when it should be multi-selectable (remember that multi-select is our new default).

Additionally, as we are now creating more entities, I've added a simple cache to the `FilterEntityFactory`. Because of this, tests that rely on the filters service should reset also the service, which fortunately it is quite easy to do.

### CODE REVIEW

File: packages/x-components/tests/e2e/cucumber/global-definitions.ts [hunk 1]
```diff
@@ -3,7 +3,15 @@ import { defineParameterType } from 'cypress-cucumber-preprocessor/steps';
...
   name: 'boolean',
   regexp: /true|false/,
-  transformer(s) {
-    return s === 'true';
+  transformer(string) {
+    return string === 'true';
+  }
+});
+
+defineParameterType({
```
Comment (javieri-empathy) [inline packages/x-components/tests/e2e/cucumber/global-definitions.ts:11]: Rollback this

File: packages/x-components/src/x-modules/facets/components/facets/__tests__/facets-provider.spec.ts [hunk 2]
```diff
@@ -129,6 +130,8 @@ function renderFacetsProviderComponent({
   stateFacets = {},
   providedFacets
 }: FacetsRenderOptions = {}): FacetsComponentAPI {
+  DefaultFacetsService.instance = new DefaultFacetsService(new FilterEntityFactory());
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets/components/facets/__tests__/facets-provider.spec.ts:133]: Just to give some context here. What it's the difference of passing the `FilterEntityFactory`?

File: packages/x-components/src/x-modules/facets/entities/filter-entity.factory.ts [hunk 3]
```diff
@@ -38,6 +38,8 @@ export class FilterEntityFactory {
    */
   protected modifiers: Record<Facet['id'], FilterEntityModifier[]> = {};
 
+  protected cache: Record<Facet['id'] | '__unknown-facet__', FilterEntity> = {};
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets/entities/filter-entity.factory.ts:41]: Please add doc.

File: packages/x-components/src/x-modules/facets/entities/filter-entity.factory.ts [hunk 4]
```diff
@@ -47,6 +49,13 @@ export class FilterEntityFactory {
    * @returns The {@link FilterEntity} created by the factory.
    */
   createFilterEntity(store: Store<RootXStoreState>, filter: Filter): FilterEntity {
+    const cacheKey = isFacetFilter(filter) ? filter.facetId : '__unknown-facet__';
+    return (
+      this.cache[cacheKey] ?? (this.cache[cacheKey] = this.instantiateFilterEntity(store, filter))
+    );
+  }
+
+  protected instantiateFilterEntity(store: Store<RootXStoreState>, filter: Filter): FilterEntity {
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets/entities/filter-entity.factory.ts:58]: Should we document this new method?

File: packages/x-components/src/x-modules/facets/service/__tests__/facets-service.spec.ts [hunk 5]
```diff
@@ -575,25 +582,57 @@ describe('testing facets service', () => {
         id: 'backend',
         facets: [newSizeFacet]
       });
+
+      expect(getSelectedFilters()).toHaveLength(1);
```
Comment (pmareke) [inline packages/x-components/src/x-modules/facets/service/__tests__/facets-service.spec.ts:586]: Should we check if the selected is the correct one (in our case the last one)?

`expect(getSelectedFilters()).toStrictEqual([newSizeFacet.filters[2]]);`

File: packages/x-components/tests/e2e/cucumber/common-steps.spec.ts [hunk 6]
```diff
@@ -214,7 +221,6 @@ Given('a results API with a known response', () => {
           ...createFilter('Muñecas', false, createFilter => [
             ...createFilter('Peluches', false),
             ...createFilter('Ropa y accesorios', false),
-            ...createFilter('Muñecas', false),
```
Comment (pmareke) [inline packages/x-components/tests/e2e/cucumber/common-steps.spec.ts:217]: Why did you remove this filter?

File: packages/x-components/tests/e2e/cucumber/multiselect-filters.feature [hunk 7]
```diff
@@ -13,18 +13,17 @@ Feature: MultiSelect filters component
...
-  # Then  selected filter is shown in the selected filters list
-  # And   search request contains selected filter
-  # TODO  https://searchbroker.atlassian.net/browse/EX-4866
-  # And   clear-filters button displays the number of selected filters
+    And   filter "<filter2>" is clicked in facet "<facetName>"
+    Then  filter "<filter1>" is shown in the selected filters list
+    And   filter "<filter2>" is shown in the selected filters list
+    And   search request contains "<filter1>" filter
+    And   search request contains "<filter2>" filter
+    And   clear-filters button should have 2 filters selected
```
Comment (pmareke) [inline packages/x-components/tests/e2e/cucumber/multiselect-filters.feature:25]: I'm not sure if I like this `2` here.

File: packages/x-components/tests/e2e/cucumber/multiselect-filters.spec.ts [hunk 8]
```diff
@@ -1,46 +1,24 @@
...
+  cy.getByDataTest('selected-filters-list')
+    .getByDataTest('filter')
+    .contains(filter)
+    .should('exist');
+});
 
-And('search request contains selected filter', function (this: { clickedFilter: string }) {
-  cy.wait('@requestWithFilter').its('request.body').should('include', this.clickedFilter);
+And('search request contains {string} filter', (filter: string) => {
+  cy.wait('@requestWithFilter').its('request.body').should('include', filter);
```
Comment (pmareke) [inline packages/x-components/tests/e2e/cucumber/multiselect-filters.spec.ts:19]: Should we use here the `json.Parse` as the other PR, I really like it.

**Review by pmareke — CHANGES_REQUESTED** [review]

File: packages/x-components/tests/e2e/cucumber/common-steps.spec.ts (hunk 6 above)
Comment (javieri-empathy) [reply packages/x-components/tests/e2e/cucumber/common-steps.spec.ts:217]: Because it was a duplicated filter which was causing an infinite loop issue. Filters are indexed by its id, and this filter has the same id  than its parent. You can imagine the rest of the story.

File: packages/x-components/src/x-modules/facets/components/facets/__tests__/facets-provider.spec.ts (hunk 2 above)
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets/components/facets/__tests__/facets-provider.spec.ts:133]: Both the facets service and the entity factory are singletons, the problem is that they are stateless singletons, which is an issue to test different flow. In this line we are "resetting" the singleton by creating new instances.

File: packages/x-components/tests/e2e/cucumber/multiselect-filters.feature (hunk 7 above)
Comment (javieri-empathy) [reply packages/x-components/tests/e2e/cucumber/multiselect-filters.feature:25]: Then think about it, and get back to this comment when you know it 🤣.

File: packages/x-components/tests/e2e/cucumber/multiselect-filters.spec.ts (hunk 8 above)
Comment (javieri-empathy) [reply packages/x-components/tests/e2e/cucumber/multiselect-filters.spec.ts:19]: I think they are both different cases. Here it would be more complicated because it is searching a filter by its label. I would leave it like this.

File: packages/x-components/src/x-modules/facets/service/__tests__/facets-service.spec.ts (hunk 5 above)
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets/service/__tests__/facets-service.spec.ts:586]: This test is just to ensure that only one filter is selected. I don't think the "which one" matters, as the outcome is that the app keeps working properly.

File: packages/x-components/tests/e2e/cucumber/multiselect-filters.feature (hunk 7 above)
Comment (pmareke) [reply packages/x-components/tests/e2e/cucumber/multiselect-filters.feature:25]: I was thinking about using a variable!

Comment (pmareke) [reply packages/x-components/tests/e2e/cucumber/multiselect-filters.feature:25]: `DON'T USE MAGIC NUMBERS`

**Review by pmareke — APPROVED** [review]

**Review by pmareke — CHANGES_REQUESTED** [review]
Fix learn errors, please.

**Review by pmareke — DISMISSED** [review]

File: packages/x-components/src/x-modules/facets/components/__tests__/clear-filters.spec.ts [hunk 9]
```diff
@@ -29,6 +30,8 @@ function renderClearFilters({
     </ClearFilters>
   `
 }: RenderFiltersOptions = {}): RenderFiltersAPI {
+  DefaultFacetsService.instance = new DefaultFacetsService(new FilterEntityFactory());
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets/components/__tests__/clear-filters.spec.ts:33]: extract this to a tests utils method

File: packages/x-components/src/x-modules/facets/entities/filter-entity.factory.ts [hunk 10]
```diff
@@ -38,15 +38,39 @@ export class FilterEntityFactory {
...
    * Creates a new FilterEntity from a filter.
    *
    * @param store - The {@link https://vuex.vuejs.org/guide/ | Vuex Store} that the entity will
    * use.
-   * @param filter - The filter necesary to know what Entity to create.
+   * @param filter - The {@link @empathyco/x-types#Filter | Filter} necessary to know what
+   * {@link FilterEntity} to create.
    * @returns The {@link FilterEntity} created by the factory.
    */
   createFilterEntity(store: Store<RootXStoreState>, filter: Filter): FilterEntity {
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets/entities/filter-entity.factory.ts:57]: this method doesn't create a FilterEntity anymore. Rename to `getFilterEntity`.

Therefore, rename the `instantiateFilterEntity` to `createFilterEntity`

File: packages/x-components/src/x-modules/facets/service/facets.service.ts [hunk 11]
```diff
@@ -83,23 +85,19 @@ export class DefaultFacetsService implements FacetsService {
...
-    const previousFilters = this.removeGroupFilters(facetsGroup.id);
+    this.removeGroupFilters(facetsGroup.id);
     facetsGroup.facets.forEach(facet => {
       this.setFacetGroup({ facetId: facet.id, groupId: facetsGroup.id });
       this.setFacet(facet);
     });
     const newFilters = facetsGroup.facets.flatMap(facet => facet.filters);
     this.setFilters(newFilters);
-    return { newFilters, previousFilters };
+    return newFilters;
```
Comment (tajespasarela) [inline packages/x-components/src/x-modules/facets/service/facets.service.ts:100]: I don't remember why we needed the `previousFilters` for.

**Review by tajespasarela — CHANGES_REQUESTED** [review]

File: packages/x-components/src/x-modules/facets/service/facets.service.ts (hunk 11 above)
Comment (javieri-empathy) [reply packages/x-components/src/x-modules/facets/service/facets.service.ts:100]: We used them to keep the previous selected state of the filters in the `updateFacets` method, but then this method was changed to first retrieve the selected filters, then replacing the filters in the store, and finally updating the filters selected state. I don't know the reason of this change, but I prefer it, as it avoids us to return 2 things from the same method. I don't think there is any issue with this approach.

File: packages/x-components/tests/e2e/cucumber/multiselect-filters.feature (hunk 7 above)
Comment (javieri-empathy) [reply packages/x-components/tests/e2e/cucumber/multiselect-filters.feature:25]: This is not a magic number, as per wikipedia, magic numbers are:

> Unique values with unexplained meaning or multiple occurrences which could (preferably) be replaced with named constants

Here I'm sure this number was perfectly understood, the human readable sentence gave enough context about it. Anyway, I've changed it just for consistency. Even if I prefer to hardcode values in tests with only one example.

**Review by tajespasarela — APPROVED** [review]

**Review by pmareke — APPROVED** [review]
